www/app/blog/2025-04-23-drop-column.mdx

```diff
@@ -0,0 +1,264 @@
+export const metadata = {
+  title: "What Really Happens When You Drop a Column in Postgres",
+  authors: ["gwenshap"],
+  image: "2025-04-23-drop-column/cover.png",
+  sizzle:
+    "When you drop a column in Postgres, the data doesn't actually go away - here's why, and what that means for you",
+  tags: ["postgres", "internals", "compliance"],
+};
+
+Dropping columns in Postgres is a common task - but few developers understand what actually happens under the hood.
+Let's dive in and explore this simple operation. We'll start with a riddle that had
+[many developers on X](https://x.com/gwenshap/status/1914751044812505441) scratching their heads. Then, we'll use some of the tools Postgres provides to
+peek behind the scenes. Finally, we'll touch on compliance implications you might not expect.
+
+If you have access to a Postgres database, you can follow along. If not, you can spin one up locally, try Nile's [test container](https://www.thenile.dev/docs/getting-started/postgres_docker), or create a database on [Nile](https://console.thenile.dev).
+
+## A Riddle
+
+Let's start with a puzzle:
+
+> We have a table with 2 columns: `id` and `name`.  
+> What will happen if we add and remove a column 2000 times?
+
+Think about it for a moment and then you can try it yourself with a simple script:
+
+```python
+import psycopg2, textwrap, sys
+conn = psycopg2.connect("postgresql://user:password@us-west-2.db.thenile.dev:5432/dbbame") ; cur = conn.cursor()
+conn.autocommit = True
+cur.execute("DROP TABLE IF EXISTS t; CREATE TABLE t(a int, b int);")
+for i in range(1, 2001):
+    cur.execute(textwrap.dedent(f"""
+        ALTER TABLE t ADD COLUMN c{i} int;
+        ALTER TABLE t DROP COLUMN c{i};
+    """))
+cur.close(); conn.close()
+```
+
+It looks harmless. The table should end up with just the original two columns. But instead, you'll hit this:
+
+```
+psycopg2.errors.TooManyColumns: tables can have at most 1600 columns
+```
+
+But... we only have two columns? What gives?
+
+## What's happening under the hood?
+
+[Postgres has a hard limit of 1600 columns per table](https://www.postgresql.org/docs/15/limits.html).
+We hit this limit when we only had two columns in our table because when you drop a column, it doesn't actually get deleted.
+
+### Why doesn't Postgres delete the column?
+
+Postgres stores table data in files, divided into 8KB pages. Each page holds metadata and the actual row data.
+
+![Diagram of a Postgres page](./2025-04-23-drop-column/pgdata.png)
```

Comment:
nit: the plain arrows are overlapping, so it's not super clear for it that the first "rowid" (which is actually called "line pointer" in postgres, and "row" is named "tuple") points to the last row and so on (even if the dotted arrow should hint it).